### DISCUSSION TITLE
Jeffries Matusita in HSI analysis

### DISCUSSION THREAD

**Discussion by iconoclast007** [community/community#52559] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

I'm trying to calculate JM distance with Python for bands in the Indian Pines data. Below is my code. Any comments? Results seem correct?

The code downloads Indian Pines and stores it in a numpy array. Calculates Bhattacharya and then uses that for Jeffries Matusita:

# Import necessary and appropriate packages

import numpy as np
import os
import pandas as pd
import requests
from [scipy.io](http://scipy.io/) import loadmat # MATlab data files
import matplotlib.pyplot as plt
from matplotlib import cm
from matplotlib.ticker import LinearLocator
from mpl_toolkits.mplot3d import Axes3D

from sklearn.metrics import normalized_mutual_info_score, mutual_info_score
from sklearn.feature_selection import mutual_info_regression

from IPython.core.display import json

def getIndianPinesData():

    if os.path.exists("Indian_pines_corrected.mat") and os.path.exists("Indian_pines_gt.mat"):
        print('files exists')
    else:
        URL = "http://www.ehu.eus/ccwintco/uploads/6/67/Indian_pines_corrected.mat"
        response = requests.get(URL)
        open("Indian_pines_corrected.mat", "wb").write(response.content)

        URL = "http://www.ehu.eus/ccwintco/uploads/c/c4/Indian_pines_gt.mat"
        response = requests.get(URL)
        open("Indian_pines_gt.mat", "wb").write(response.content)
    
    X = loadmat('Indian_pines_corrected.mat')['indian_pines_corrected']
    y = loadmat('Indian_pines_gt.mat')['indian_pines_gt']

    print(type(X),type(y))
    print("Dimensions of Indian Pines array are:",X.shape,", with",X.shape[2],'bands')
    
    
    return X,y

def computeBhattacharyya(bands):
    numberOfBands = bands.shape[2]
    BHdistances = np.zeros((numberOfBands, numberOfBands))
    for i in range(numberOfBands):
        for j in range(numberOfBands):
            # Get the two bands
            band1 = bands[:,:,i]
            band2 = bands[:,:,j]
        
            # Calculate the means and standard deviations of the two bands
            mean1 = np.mean(band1)
            mean2 = np.mean(band2)
            std1 = np.std(band1)
            std2 = np.std(band2)
        
            # Calculate the Bhattacharyya distance
            b_dist = np.sqrt(1 - np.exp(-0.25 * (mean1 - mean2)**2 / (std1**2 + std2**2)))
        
            # Store the distance in the distances matrix
            BHdistances[i,j] = b_dist
   
    print("Bhattacharyya distance matrix\n")
    print("Bhattacharyya distance matrix:\n\n",BHdistances)
    
    im = plt.imshow(BHdistances, cmap='hot')
    plt.colorbar(im, orientation='horizontal')
    plt.show()
    return BHdistances

def computeJM(BH):
    
    numberOfBands = BH.shape[1]
    JM = np.zeros((numberOfBands, numberOfBands))
    '''
    The Jeffries-Matusita (JM) measure is a statistical measure of the similarity between two 
    multivariate distributions, and it can be calculated using the Bhattacharyya distance matrix. 
    The formula for calculating the JM measure from the Bhattacharyya distance matrix is:

    JM = sqrt(2 * (1 - exp(-BD)))

    where BD is the Bhattacharyya distance matrix.
    '''
    for i in range(numberOfBands):
        for j in range(numberOfBands):
            # Get the two bands
            if i != j:
                JM[i][j] = np.sqrt(2 * (1 - np.exp(-BH[i][j])))
    print("plot Jeffries Matusita measurement\n")
    print(JM)
    
    im = plt.imshow(JM, cmap='hot')
    plt.colorbar(im, orientation='horizontal')
    plt.show()

def main():
    #loadHSIdatafile()
    #checkFile()
    A, b = getIndianPinesData()
    BHdistances = computeBhattacharyya(A)
    print(BHdistances.shape)
    computeJM(BHdistances)

if __name__ == '__main__':
    main()

%%%%%%%%%%%%%%%%%
%
Does this look correct?

**Comment by github-actions** [comment]
🕒 **Discussion Activity Reminder** 🕒

This Discussion has been labeled as dormant by an automated system for having no activity in the last 60 days. Please consider one the following actions:

1️⃣ Close as Out of Date: If the topic is no longer relevant, close the Discussion as `out of date` at the bottom of the page.

2️⃣ Provide More Information: Share additional details or context — or let the community know if you've found a solution on your own.

3️⃣ Mark a Reply as Answer: If your question has been answered by a reply, mark the most helpful reply as the solution.

Note: This dormant notification will only apply to Discussions with the `Question` label. To learn more, see our [recent announcement](https://github.com/orgs/community/discussions/70478).

Thank you for helping bring this Discussion to a resolution! 💬
